middle_end/flambda/to_cmm/un_cps_closure.ml

```diff
@@ -105,40 +113,73 @@ let order_env_vars env l acc =
             Var_within_closure.print env_var
     ) acc l
 
-let order env closures env_vars =
-  order_env_vars env env_vars
-    (order_closures env closures Numbers.Int.Map.empty)
-
-let rec layout_aux map acc i =
-  match Numbers.Int.Map.find_first (fun j -> j >= i) map with
-  | exception Not_found -> List.rev acc
-  | (j, (Closure _ as slot)) when j = 0 ->
-      assert (i = j);
-      assert (acc = []);
-      layout_aux map [(0, slot)] (i + 1)
-  | (j, (Closure _ as slot)) ->
-      layout_aux map ((j, slot) :: (j - 1, Infix_header) :: acc) (j + 1)
-  | (j, (Env_var _ as slot)) ->
-      layout_aux map ((j, slot) :: acc) (j + 1)
-  | (_, Infix_header) ->
+let layout_aux j slot (startenv, acc_slots) =
+  match slot with
+  | Closure _ when j = 0 ->
+    assert (acc_slots = []);
+    assert (startenv = None);
+    let acc_slots = [(0, slot)] in
+    startenv, acc_slots
+  | Closure _ ->
+    assert (startenv = None);
+    let acc_slots = (j, slot) :: (j - 1, Infix_header) :: acc_slots in
+    startenv, acc_slots
+  | Env_var _ ->
+    let startenv =
+      match startenv with
+      | Some i -> assert (i < j); startenv
+      | None -> Some j
+    in
+    let acc_slots = (j, slot) :: acc_slots in
+    startenv, acc_slots
+  | Infix_header ->
       (* Internal invariant: such layout slots are not generated by
          the {order} function, so they should not appear. *)
       assert false
 
 let layout env closures env_vars =
-  let map = order env closures env_vars in
-  layout_aux map [] 0
+  let map = Numbers.Int.Map.empty in
+  let map = order_env_vars env env_vars map in
+  let map = order_closures env closures map in
```

Comment:
This feels a bit odd to me (in particular the first definition of `map` to `Numbers.Int.Map.empty`).
I would be in favour of a single let binding, using either a single multi-line application (like the one that used to be in the `order` function) or pipe applications (`|>`), whichever feels more readable to you.

Reply:
Switched to using the pipe operator.